### PR TITLE
Fix #3089 Cancel NSURLSessionStreamTask could cause a crash in iOS 10

### PR DESCRIPTION
**Pull request by pasin**
- Instead of cancel, close read/write socket instead for the case of an error occurred to avoid the crash in iOS 10.

- In readClosedForStreamTask: and writeClosedForStreamTask:, also validate the task if the task is still the current task. This is required because we canceled the current task is in - receivedHTTPResponse due to a redirect or auth challenge. In iOS 10 (not iOS 11), after canceling the task, the -readClosedForStreamTask: and -writeClosedForStreamTask: delegate will be called. Without checking the -streamClosed: method will be called, and there could be a race condition which causes the new task to be canceled in the -streamClosed: method. In general we don't want -streamClosed: to be called for this case.

- We should either close the sockets or cancel the task in didCloseWithCode:reason: as well otherwise the sockets will leak. To be safe, I close the sockets instead of canceling the task.

- Enabled testStopReplicatorAfterOffline_SG and ensure that the replicator will be stopped after running the replicator even though the test failed.

#2089

### CODE REVIEW

File: Objective-C/Internal/Replicator/CBLWebSocket.mm [hunk 1]
```diff
@@ -457,14 +456,18 @@ - (void) URLSession: (NSURLSession *)session
 
 - (void)URLSession:(NSURLSession *)session readClosedForStreamTask:(NSURLSessionStreamTask *)task
 {
-    [self streamClosed: NO];
+    if (task == _task)
```
Comment (snej) [inline Objective-C/Internal/Replicator/CBLWebSocket.mm]: It'd be cleaner to move this test (and the one in the next method) into -streamClosed:, since we're trying to share the logic for both the read and write streams.

File: Objective-C/Internal/Replicator/CBLWebSocket.mm [hunk 2]
```diff
@@ -538,6 +543,12 @@ - (void) didCloseWithError: (NSError*)error {
 }
 
 
+- (void) closeTask {
+    [_task closeWrite];
+    [_task closeRead];
+}
```
Comment (snej) [inline Objective-C/Internal/Replicator/CBLWebSocket.mm]: Should you set `_task = nil` here, since it's now closed? Then you don't have to set it to nil after the various calls to -closeTask.

Comment (pasin) [reply Objective-C/Internal/Replicator/CBLWebSocket.mm]: I tried to do this, but the changing scope is getting bigger. There is some complication especially with -didCloseWithError: that shouldn't be called when _task is nil. I think we should do this in post 2.0 after we have enough time to test the change to cover all edge cases. I put todo in the code.
